module: add module.stripTypeScriptTypes

marco-ippolito opened the pull request:
This PR introduces a new api in `node:module`.

Fixes: https://github.com/nodejs/node/issues/54300

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -107,6 +110,16 @@ changes:
     experimental modules API. We do not recommend using it in a production
     environment. For detailed information, see
     [Support of dynamic `import()` in compilation APIs][].
+  * `transform` {Object} An object containing options for type stripping.
```
Could this be called `'typescript'`? Similar to `process.features.typescript`?

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -107,6 +110,16 @@ changes:
     experimental modules API. We do not recommend using it in a production
     environment. For detailed information, see
     [Support of dynamic `import()` in compilation APIs][].
+  * `transform` {Object} An object containing options for type stripping.
+    Only available when the `--experimental-strip-types` flag is enabled.
+    * `mode` {string} The mode of type stripping. Possible values are:
+      * `strip-only`: Strip all types, without transforming TypeScript only features,
+        for more information read [type-stripping][].
+      * `transform`: Transforms TypeScript features, for info read [transform TypeScript features][].
+    * `sourceMap` {boolean} Generate source maps for the transformed code, if `mode` is `transform`,
+      otherwise it is not necessary.
+      The filename for the source map is the same as the filename of the script.
+      **Default:** `false`.
```
```suggestion
	* `transform` {Object} An object that holds options for type stripping. This is accessible only when the `--experimental-strip-types` flag is enabled.
	  * `mode` {string} Specifies the type stripping mode. The possible options are:
	    * `strip-only`: Removes all types without altering TypeScript-specific features. For more details, refer to [type-stripping][].
	    * `transform`: Modifies TypeScript features. For additional information, see [transform TypeScript features][].
	  * `sourceMap` {boolean} If `mode` is set to `transform`, generates source maps for the transformed code; otherwise, it is unnecessary. The source map will have the same filename as the script.  
	  **Default:** `false`.
```

marco-ippolito replied:
What is the change?

RedYetiDev replied:
It's just worded a bit different. IMO the change makes it more understandable, but feel free to ignore:

E.G.:
``An object containing options for type stripping.
    Only available when the `--experimental-strip-types` flag is enabled.`` becomes ``An object that holds options for type stripping. This is accessible only when the `--experimental-strip-types` flag is enabled.``

marco-ippolito replied:
This is a draft PR why are you nitpicking on wording?

RedYetiDev replied:
Sorry, just my two-cents, I'll leave it be.

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -1044,6 +1060,16 @@ changes:
   * `contextExtensions` {Object\[]} An array containing a collection of context
     extensions (objects wrapping the current scope) to be applied while
     compiling. **Default:** `[]`.
+  * `transform` {Object} An object containing options for type stripping.
+    Only available when the `--experimental-strip-types` flag is enabled.
+    * `mode` {string} The mode of type stripping. Possible values are:
+      * `strip-only`: Strip all types, without transforming TypeScript only features,
+        for more information read [type-stripping][].
+      * `transform`: Transforms TypeScript features, for info read [transform TypeScript features][].
+    * `sourceMap` {boolean} Generate source maps for the transformed code, if `mode` is `transform`,
+      otherwise it is not necessary.
+      The filename for the source map is the same as the filename of the script.
+      **Default:** `false`.
```
Ditto

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -1335,6 +1364,16 @@ changes:
     experimental modules API. We do not recommend using it in a production
     environment. For detailed information, see
     [Support of dynamic `import()` in compilation APIs][].
+  * `transform` {Object} An object containing options for type stripping.
+    Only available when the `--experimental-strip-types` flag is enabled.
+    * `mode` {string} The mode of type stripping. Possible values are:
+      * `strip-only`: Strip all types, without transforming TypeScript only features,
+        for more information read [type-stripping][].
+      * `transform`: Transforms TypeScript features, for info read [transform TypeScript features][].
+    * `sourceMap` {boolean} Generate source maps for the transformed code, if `mode` is `transform`,
+      otherwise it is not necessary.
+      The filename for the source map is the same as the filename of the script.
+      **Default:** `false`.
```
Ditto

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -1443,6 +1485,16 @@ changes:
     scheduled through `Promise`s and `async function`s) will be run immediately
     after the script has run. They are included in the `timeout` and
     `breakOnSigint` scopes in that case.
+  * `transform` {Object} An object containing options for type stripping.
+    Only available when the `--experimental-strip-types` flag is enabled.
+    * `mode` {string} The mode of type stripping. Possible values are:
+      * `strip-only`: Strip all types, without transforming TypeScript only features,
+        for more information read [type-stripping][].
+      * `transform`: Transforms TypeScript features, for info read [transform TypeScript features][].
+    * `sourceMap` {boolean} Generate source maps for the transformed code, if `mode` is `transform`,
+      otherwise it is not necessary.
+      The filename for the source map is the same as the filename of the script.
+      **Default:** `false`.
```
Ditto

RedYetiDev commented on doc/api/vm.md:
```diff
@@ -1532,6 +1587,16 @@ changes:
     experimental modules API. We do not recommend using it in a production
     environment. For detailed information, see
     [Support of dynamic `import()` in compilation APIs][].
+  * `transform` {Object} An object containing options for type stripping.
+    Only available when the `--experimental-strip-types` flag is enabled.
+    * `mode` {string} The mode of type stripping. Possible values are:
+      * `strip-only`: Strip all types, without transforming TypeScript only features,
+        for more information read [type-stripping][].
+      * `transform`: Transforms TypeScript features, for info read [transform TypeScript features][].
+    * `sourceMap` {boolean} Generate source maps for the transformed code, if `mode` is `transform`,
+      otherwise it is not necessary.
+      The filename for the source map is the same as the filename of the script.
+      **Default:** `false`.
```
Ditto

RedYetiDev commented on lib/internal/modules/helpers.js:
```diff
@@ -324,35 +324,42 @@ let typeScriptParser;
+ * @returns {string} The TypeScript parsing mode
  */
-let sourceMapEnabled;
+function getTypeScriptParsingMode() {
+  if (typeScriptParsingMode) {
+    return typeScriptParsingMode;
+  }
+  typeScriptParsingMode = getOptionValue('--experimental-transform-types') ? 'transform' : 'strip-only';
+  return typeScriptParsingMode;
+}
```
```suggestion
const typeScriptParsingMode = getLazy(() => getOptionValue('--experimental-transform-types') ? 'transform' : 'strip-only');
```

RedYetiDev commented on lib/internal/modules/helpers.js:
```diff
@@ -324,35 +324,42 @@ let typeScriptParser;
-    const amaro = require('internal/deps/amaro/dist/index');
-    // Default option for Amaro is to perform Type Stripping only.
-    typeScriptParsingMode = getOptionValue('--experimental-transform-types') ? 'transform' : 'strip-only';
-    sourceMapEnabled = getOptionValue('--enable-source-maps');
-    // Curry the transformSync function with the default options.
-    typeScriptParser = amaro.transformSync;
-  }
+  const amaro = require('internal/deps/amaro/dist/index');
+  typeScriptParser = amaro.transformSync;
   return typeScriptParser;
```
Ditto w/ `getLazy`

joyeecheung commented:
I think it makes more sense for it be just in its own API as a function, which can be used to strip the type before the text is passed into any of the `vm.Script`/`vm.SourceTextModule`/`vm.compileFunction` APIs. That'll also be useful for those who want to transform the text but not compile it at all.

devsnek commented:
there's also #54250

marco-ippolito commented:
> there's also #54250

unfortunately it seems have stalled for a while.
Anyways good to review, I'm just not sure whether to be strict and throw when mode is `strip-types` and `sourceMap` is passed, otherwise will be ignored and should be counter intuitive

atlowChemi commented on lib/vm.js:
```diff
@@ -400,6 +403,45 @@ const vmConstants = {
 
 ObjectFreeze(vmConstants);
 
+function stripTypeScriptTypes(code, options = {}) {
```
```suggestion
function stripTypeScriptTypes(code, options = kEmptyObject) {
```

atlowChemi commented on lib/vm.js:
```diff
@@ -400,6 +403,45 @@ const vmConstants = {
 
+function stripTypeScriptTypes(code, options = {}) {
+  emitExperimentalWarning('vm.stripTypeScriptTypes');
+  const { mode = 'strip-only', sourceMap = false, filename = '' } = options;
+  validateObject(options, 'options');
```
The order seems wrong. if `options` was provided with any non-object value other than `undefined` we should throw an `ERR_INVALID_ARG_TYPE`, but this currently could potentially throw an error such as `can't destructure property of options as it is null`

```suggestion
  validateObject(options, 'options');
  const { mode = 'strip-only', sourceMap = false, filename = '' } = options;
```

atlowChemi commented on lib/vm.js:
```diff
@@ -400,6 +403,45 @@ const vmConstants = {
+  emitExperimentalWarning('vm.stripTypeScriptTypes');
+  const { mode = 'strip-only', sourceMap = false, filename = '' } = options;
+  validateObject(options, 'options');
+  validateString(mode, 'options.mode');
+  if (mode !== 'strip-only' && mode !== 'transform') {
+    throw new ERR_INVALID_ARG_VALUE('options.mode', mode, 'must be either ' +
+                                    '"strip-only" or "transform"');
+  }
```
```suggestion
  validateOneOf(mode, 'options.mode', ['strip-only', 'transform']);
```

atlowChemi commented on lib/vm.js:
```diff
@@ -400,6 +403,45 @@ const vmConstants = {
+  }
+
+  if (mode === 'strip-only') {
+    if (sourceMap)
+      throw new ERR_INVALID_ARG_VALUE('options.sourceMap', mode, 'must be either ' +
+      'false or undefined when mode is "strip-only"');
+    if (filename)
+      throw new ERR_INVALID_ARG_VALUE('options.filename', mode, 'must be either ' +
+      'undefined or an empty string when mode is "strip-only"');
```
consider using `validateOneOf` (though it won't have the suffix of `when mode is "strip-only"`)
```suggestion
    validateOneOf(sourceMap, 'options.sourceMap', [false, undefined]);
    validateOneOf(filename, 'options.filename', ['', undefined]);
```

legendecas commented:
I think this could be benefited from a dedicated module `parser` (name bikeshed welcome), instead of `vm` since the updated PR doesn't involve with any code evaluation. This module could provide additional helpers to address similar needs on manipulating JS source codes.

joyeecheung commented:
Maybe putting it on module also makes sense? Or module.typescript if we can foresee adding other typescript-related utilities in the future

marco-ippolito commented:
> I think this could be benefited from a dedicated module `parser` (name bikeshed welcome), instead of `vm` since the updated PR doesn't involve with any code evaluation. This module could provide additional helpers to address similar needs on manipulating JS source codes.

I'm ok with a new module, I think parser is ok, I'd like to hear more opinions about it

marco-ippolito commented:
If we decide to ship it I'll drop the vm commits

joyeecheung commented:
If this needs to be a new module, then it should be added to https://github.com/nodejs/node/blob/b4e8f1b6bb3616ba222c4513218aa1fa9d543d8e/lib/internal/bootstrap/realm.js#L130 to block `require('parser')`, otherwise it breaks https://www.npmjs.com/package/parser or if someone creates an alias of `parser` in node_modules.

devsnek commented:
i think if the goal is to expose a generic ts api, then something from npm should be used. if the goal is to expose "handle ts modules the way the current node process does", it should be exported from `node:module`.

marco-ippolito commented:
Sounds reasonable, I also refactored the typescript internals in its own file.
The only thing that upsets me is that we have now an `internalStripTypeScriptTypes` and `stripTypeScriptTypes `.
I'd appreciate some help with naming 🙏🏼

legendecas commented:
I believe the API exposed was not limited to "handle ts modules the way the current node process does": it allows `transform` mode even if the process is in `strip` mode or, without ts-support enabled. So IMO this lies more in "generic" API.

marco-ippolito commented:
I dont have a strong opinion on wheter should go on a new module or in the `node:module` one.
Maybe create a new module when we have more than 1 api? If its experimental it can be moved without going through a deprecation cycle.

atlowChemi commented on lib/internal/modules/typescript.js:
```diff
@@ -0,0 +1,106 @@
+ * @returns {TransformOutput} The stripped TypeScript code.
+ */
+function stripTypeScriptModuleTypes(source, filename) {
+  assert(typeof source === 'string');
+  const options = {
+    __proto__: null,
+    mode: getTypeScriptParsingMode(),
+    sourceMap: getOptionValue('--enable-source-maps'),
+    filename,
+  };
+  const { code, map } = parseTypeScript(source, options);
+  if (map) {
+    // TODO(@marco-ippolito) When Buffer.transcode supports utf8 to
+    // base64 transformation, we should change this line.
+    const base64SourceMap = Buffer.from(map).toString('base64');
+    return `${code}\n\n//# sourceMappingURL=data:application/json;base64,${base64SourceMap}`;
+  }
+  // Source map is not necessary in strip-only mode. However, to map the source
+  // file in debuggers to the original TypeScript source, add a sourceURL magic
+  // comment to hint that it is a generated source.
+  return `${code}\n\n//# sourceURL=${filename}`;
```
Can't we reduce duplication by reusing `stripTypeScriptTypes`?

```suggestion
  assert(typeof source === 'string');
  const mode = getTypeScriptParsingMode();
  const options = {
    __proto__: null,
    mode,
    sourceMap: getOptionValue('--enable-source-maps'),
    filename,
  };
  const result = stripTypeScriptTypes(source, options);
  // Source map is not necessary in strip-only mode. However, to map the source
  // file in debuggers to the original TypeScript source, add a sourceURL magic
  // comment to hint that it is a generated source.
  if (mode === 'strip-only') {
    return `${result}\n\n//# sourceURL=${filename}`;
  }
```

marco-ippolito replied:
It would perform unnecessary validations, there a few things that can be factorized like adding the sourcemap but other than that they are different

legendecas commented:
To be honest, I don't feel `node:module` is a good place for such utilities as the function don't involve with the module loading / affecting the module system, like in cases like `vm.compileFunction`.

marco-ippolito commented:
I see there is some disagreement around.
After playing around a bit I think creating a new module is the right thing.
The reason is:
- It cannot stay in vm since it doesnt run code
- It should stay in module since its not really related to a particular module system. Its unrelate to `--experimental-strip-types` since it does not depend on the flag

atlowChemi commented on test/parallel/test-module-strip-types.js:
@marco-ippolito shouldn't this file be renamed to `test-parser-strip-types`?

atlowChemi commented on doc/api/parser.md:
````diff
@@ -0,0 +1,70 @@
+
+The `node:parser` module provides utilities for working with source code. It can be accessed using:
+
+```js
+const parser = require('node:parser');
+```
````
````suggestion
```mjs
import parser from 'node:parser';
```
```cjs
const parser = require('node:parser');
```
````

atlowChemi commented on doc/api/parser.md:
````diff
@@ -0,0 +1,70 @@
+  When mode is `'strip-only'`, source maps are not generated, because locations are preserved.
+  If `sourceMap` or `filename` is provided, when mode is `'strip-only'`, an error will be thrown.
+
+```js
+const parser = require('node:parser');
+const code = `const a: number = 1;`;
+const strippedCode = parser.stripTypeScriptTypes(code);
+console.log(strippedCode);
+// Prints: const a         = 1;
+```
````
````suggestion
```mjs
import parser from 'node:parser';

const code = `const a: number = 1;`;
const strippedCode = parser.stripTypeScriptTypes(code);
console.log(strippedCode);
// Prints: const a         = 1;
```
```cjs
const parser = require('node:parser');
const code = `const a: number = 1;`;
const strippedCode = parser.stripTypeScriptTypes(code);
console.log(strippedCode);
// Prints: const a         = 1;
```
````

atlowChemi commented on doc/api/parser.md:
````diff
@@ -0,0 +1,70 @@
+
+When `mode` is `'transform'`, the code is transformed to JavaScript:
+
+```js
+const parser = require('node:parser');
+const code = `
+  namespace MathUtil {
+    export const add = (a: number, b: number) => a + b;
+  }`;
+const strippedCode = parser.stripTypeScriptTypes(code, { mode: 'transform', sourceMap: true });
+console.log(strippedCode);
+// Prints:
+// var MathUtil;
+// (function(MathUtil) {
+//     MathUtil.add = (a, b)=>a + b;
+// })(MathUtil || (MathUtil = {}));
+//# sourceMappingURL=data:application/json;base64, ...
+```
````
````suggestion
```mjs
import parser from 'node:parser';

const code = `
  namespace MathUtil {
    export const add = (a: number, b: number) => a + b;
  }`;
const strippedCode = parser.stripTypeScriptTypes(code, { mode: 'transform', sourceMap: true });
console.log(strippedCode);
// Prints:
// var MathUtil;
// (function(MathUtil) {
//     MathUtil.add = (a, b)=>a + b;
// })(MathUtil || (MathUtil = {}));
//# sourceMappingURL=data:application/json;base64, ...
```
```cjs
const parser = require('node:parser');
const code = `
  namespace MathUtil {
    export const add = (a: number, b: number) => a + b;
  }`;
const strippedCode = parser.stripTypeScriptTypes(code, { mode: 'transform', sourceMap: true });
console.log(strippedCode);
// Prints:
// var MathUtil;
// (function(MathUtil) {
//     MathUtil.add = (a, b)=>a + b;
// })(MathUtil || (MathUtil = {}));
//# sourceMappingURL=data:application/json;base64, ...
```
````

legendecas reviewed (commented):
LGTM with a new experimental module to expose source code manipulation operations.

richardlau commented:
> * It should not live in module since its not really related to a particular module system. Its unrelated to `--experimental-strip-types` since it does not depend on the flag

It does, however, depend on amaro being available (not the case when configured `--without-amaro`).

atlowChemi commented on lib/internal/parser/typescript.js:
```diff
@@ -0,0 +1,125 @@
+ * @typedef {object} TransformOptions
+ * @property {string} mode The TypeScript parsing mode, either 'strip-only' or 'transform'.
+ * @property {boolean} sourceMap Whether to generate source-maps.
+ * @property {string} filename The filename of the source code.
```
```suggestion
 * @property {string} sourceUrl The filename of the source code.
```

RedYetiDev commented on doc/api/parser.md:
````diff
@@ -0,0 +1,98 @@
+const parser = require('node:parser');
+```
+
+## `parser.stripTypeScriptTypes(code[, options])`
````
Nit: imo the name is a bit misleading, as this does more than just strip types?

anonrig commented:
> I see there is some disagreement about where this feature should go.
> 
> After playing around a bit I think creating a new module is the right thing.
> 
> The reason is:
> 
> - It should not live in vm since it doesnt run code
> 
> - It should not live in module since its not really related to a particular module system. Its unrelated to `--experimental-strip-types` since it does not depend on the flag

@marco-ippolito do you see a different parser added into this new module soon? I don't think we should add a new module just for a single function. Why don't we put it under util?

marco-ippolito commented:
@anonrig  I think there is the possibility of adding more functions to manipulate code. But imho this new feature does not belong in any of the current modules

anonrig commented:
@marco-ippolito We have parseEnv file in node:util. So this means we will add that function to node:parser as well?

marco-ippolito commented:
> @marco-ippolito We have parseEnv file in node:util. So this means we will add that function to node:parser as well?

Yes

anonrig reviewed (changes requested):
I'm sorry for doing this, but I think we shouldn't introduce a new module just for a single method. This should live in node:util with similar functions like parseEnv.

marco-ippolito commented:
I feel like I motivated the decision in previous comment, that function that maniupulate source code without evaluating should be moved to this module, like `parseEnv`, not the opposite. Everything can go in util.

mcollina commented:
@marco-ippolito I think this is a textbook case for calling a vote.

I dislike the use of name `node:parser`, something like `node:code_transformations` would fit better or even as an utility in `node:module`.

`util` is fine, as long as the dependency is loaded dynamically.

marco-ippolito commented:
I moved it to the `module:` module (lol)
I moved all the typescript related stuff in a separate file because it was bloating the `module/helpers` file. (everything is a util and a helper).
Hope this reduces the disagreements around this feature.

marco-ippolito commented:
I screwed the rebase

jasnell commented on test/parallel/test-module-strip-types.js:
```diff
@@ -0,0 +1,92 @@
+
+const common = require('../common');
+if (!process.config.variables.node_use_amaro) common.skip('Requires Amaro');
+const assert = require('assert');
```
Non-blocking: for consistency

```suggestion
const assert = require('node:assert');
```

joyeecheung commented on lib/internal/modules/typescript.js:
```diff
@@ -0,0 +1,147 @@
+ * @returns {string} The code with the source map attached.
+ */
+function addSourceMap(code, sourceMap) {
+  // TODO(@marco-ippolito) When Buffer.transcode supports utf8 to
```
I know it's already there, but FYI you could just use `internalBinding('buffer').btoa(code))` and avoid the intermediate buffer.

marco-ippolito replied:
Done thanks

mhdawson commented:
Looks like this is moving forward, removing TSC agenda lable after discussion in todays meeting, please re-add if there is still something the TSC needs to discuss

marco-ippolito commented:
> Looks like this is moving forward, removing TSC agenda lable after discussion in todays meeting, please re-add if there is still something the TSC needs to discuss

its still blocked by @anonrig review

anonrig commented:
> its still blocked by @anonrig review

I propose having a vote on this. @nodejs/tsc

marco-ippolito commented:
> > its still blocked by @anonrig review
> 
> I propose having a vote on this. @nodejs/tsc

Fine have it your way.

anonrig reviewed (approved):
It was an error on my side. I didn't get the previous notifications and assumed that we're still creating a new module.

RedYetiDev commented:
Hey, I've read over the past few messages, and based on them, I've adjusted the labels of this issue, LMK if that was a mistake, as this was on the agenda for TSC, but per the comments above, I've removed it.

ShogunPanda reviewed (approved):
LGTM!

marco-ippolito commented:
`Downloading patch for 55282fatal: couldn't find remote ref refs/pull/55282/merge`
same error when trying manually with `git node land`

richardlau commented:
> `Downloading patch for 55282fatal: couldn't find remote ref refs/pull/55282/merge` same error when trying manually with `git node land`
> 
> Let me try rebase and see if it makes difference

FWIW Prior to the rebase I was seeing:
```console
$ git ls-remote upstream refs/pull/55282/*
c1c262a134f21bb350e0dfe4c8c50e2009195a04        refs/pull/55282/head
$
```
which was consistent with the error (no `refs/pull/55282/merge`). After the rebase:
```console
$ git ls-remote upstream refs/pull/55282/*
8532f1e1e07538901a102c1880700f767d0e6686        refs/pull/55282/head
efb62871253f691abd0ac770e7361f20af13d1d2        refs/pull/55282/merge
$
```
🤞

ruyadorno commented:
This commit does not land cleanly on `v22.x-staging` and will need manual backport in case we want it in **v22.x**.
